---
title: 'Harmonization of Proteomics Data'
author: "Vivien Wiltzscht & Klaudia Adamowicz"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    number_sections: false
    toc: true
    toc_float: true
    toc_depth: 5
---

```{r setup, include=FALSE}
reticulate::use_virtualenv("/home/python_env", required = TRUE)
library("proharmed")
```

## Run Proharmed Harmonization

In this guide, we'll navigate through the essential steps of harmonizing proteomics data. Our focus is on refining proteomic datasets to enhance their clarity and utility. We begin with filtering protein IDs to ensure data accuracy, followed by gene name remapping and reduction for streamlined dataset representation. We also explore the mapping of orthologs, crucial for cross-species analysis. Utilizing the `proharmed` package, we'll demonstrate these processes and complement them with insightful visualizations, providing a clear view of each step in the data processing journey. This guide is a concise resource for effective proteomics data harmonization.

### Config

```{r config}
prot_file <- "report.pg_matrix.tsv"
meta_file <- "meta_data.csv"

save_plots <- TRUE
# Define the final output directory or set NULL if results should note be saved
final_out_dir <- "harmonization_results"
# Creates the folder if it doesn't exist
if(!is.null(final_out_dir)){dir.create(final_out_dir, recursive = TRUE, showWarnings = FALSE)}
```

### Load Data & Function

```{r load_data}
# Load data
prot_data <- data.table::fread(prot_file)
meta_data <- data.table::fread(meta_file)
```


```{r helper_function, include=FALSE}
# Function to save each data.frame in the list
save_data_frames <- function(in_list, subdir) {
  for (name in names(in_list)) {
    filepath <- file.path(subdir, paste0(name, ".csv"))
    # Create a copy of the data.frame for modification
    df <- in_list[[name]]
    # Check for columns that are lists and concatenate their elements
    list_cols <- sapply(df, is.list)
    if (any(list_cols)) {
      df[list_cols] <- lapply(df[list_cols], function(col) {
        sapply(col, function(cell) {
          if (is.null(cell)) return(NA)
          paste(cell, collapse = ";")
        })
      })
    }
    write.csv(df, file = filepath, row.names = FALSE)
  }
}
```


### Filtering Protein IDs

The `filter_protein_ids` function is designed to process a DataFrame containing protein IDs. It allows for the removal of decoy IDs (such as those starting with "REV__" or "CON__"), contamination IDs, and the option to filter IDs based on a specified organism. Additionally, the function can restrict protein IDs to reviewed ones and manage empty cells in the dataset.

#### Parameters

- `data`: A DataFrame that includes a column with protein IDs.
- `protein_column`: The name of the column in `data` that contains the protein IDs.
- `organism`: (Optional) Specify the organism to filter the protein IDs. [human, rat, mouse, rabbit]
- `rev_con`: (Optional) A boolean value indicating if decoy protein IDs should be kept.
- `keep_empty`: A boolean value indicating if rows with empty protein ID cells should be retained.
- `res_column`: The name of the column where the filtered protein IDs will be stored. If `NULL`, the original `protein_column` will be overwritten.
- `reviewed`: (Optional) A boolean value indicating if the protein IDs should be reduced to reviewed ones only.

#### Run Method

```{r filter_proteins}
# Define the parameters
protein_column <- "Protein.Ids"
organism <- "rat"  
rev_con <- FALSE 
keep_empty <- TRUE
res_column <- "FilteredProteinIDs"
reviewed <- FALSE

# Call the function
filtered_prot_results <- proharmed::filter_protein_ids(data = prot_data, protein_column = protein_column, 
                                                       organism = organism, rev_con = rev_con, 
                                                       keep_empty = keep_empty, res_column = res_column,
                                                       reviewed = reviewed)

filtered_prot_data <- filtered_prot_results$Modified_Data
```

#### Save Dataframes

```{r filter_proteins_save, include=FALSE}
# Create a subdirectory for filtered_protein_ids
if(!is.null(final_out_dir)){
  subdir <- file.path(final_out_dir, "filtered_protein_ids", "")
  dir.create(subdir, recursive = TRUE, showWarnings = FALSE)
  # Save the data.frames in the list to the specified subdirectory
  save_data_frames(filtered_prot_results, subdir)
}
```


#### Visualizing results

##### Overview Plot

`create_overview_plot` generates an overview plot of the logging data. This plot provides a high-level summary of the data processing steps.

Parameters:

- `logging`: A DataFrame containing the overview logging data returned by one of the proharmed methods.
- `out_dir`: (Optional) The directory where the plot will be saved.
- `file_type`: (Optional) The file format for the saved plot (e.g., png, pdf, jpg).

Returns: A list of two `ggplot` objects.

##### Detailed Plot for Filtered Protein IDs

`create_filter_detailed_plot` creates a detailed plot specifically for the logging data of the `filter_protein_ids` method. This plot provides insights into how the protein IDs were filtered based on various criteria.

Parameters:

- `logging`: A DataFrame containing detailed logging data returned by the `filter_protein_ids` method.
- `organism`: The organism specified in the `filter_protein_ids` method.
- `reviewed`: The reviewed parameter value used in the `filter_protein_ids` method.
- `decoy`: The rev_con parameter value used in the `filter_protein_ids` method.
- `out_dir`: (Optional) The directory where the plot will be saved.
- `file_type`: (Optional) The file format for the saved plot.

Returns: A `ggplot` object.

```{r filter_proteins_plots, message=FALSE, warning=FALSE}
if(save_plots && !is.null(final_out_dir))
  { out_dir=paste0(subdir, "/") } else { out_dir=NULL }

file_type="png"

# Generate and visualize the overview plot
overview_plots <- proharmed::create_overview_plot(logging = filtered_prot_results$Overview_Log, 
                                                  out_dir = out_dir, file_type = file_type)
print(overview_plots[[1]]) # Distribution of number of entries per line
print(overview_plots[[2]]) # Number of entries added and removed

# Generate and visualize the detailed filter plot
detailed_plot <- proharmed::create_filter_detailed_plot(logging = filtered_prot_results$Detailed_Log, 
                                                        organism = organism, reviewed = reviewed, 
                                                        decoy = rev_con, out_dir = out_dir, file_type = file_type)
print(detailed_plot) # Number of entries removed by cause
```

### Remap Gene Names

The `remap_genenames` function in the proharmed package is designed to remap protein IDs to their associated gene names, offering various modes for filling empty entries or optionally replacing existing gene names.

Modes:

- `all`: Uses primarily information from the FASTA headers, supplemented with UniProt information.
- `fasta`: Utilizes information extracted from FASTA headers.
- `uniprot`: Employs mapping information from UniProt to use all gene names.
- `uniprot_primary`: Uses UniProt mapping information but limits to primary gene names.
- `uniprot_one`: Uses UniProt mapping but only employs the most frequent single gene name.

#### Parameters

- `data`: A DataFrame containing a column with protein IDs.
- `mode`: Mode of refilling (options are all, fasta, uniprot, uniprot_primary, uniprot_one).
- `protein_column`: Name of the column with protein IDs.
- `gene_column`: (Optional) Name of the column with gene names. Defaults to "Gene Names".
- `res_column`: Name of the column for results. If NULL, the `gene_column` will be overwritten.
- `skip_filled`: (Optional) A boolean value indicating if already filled gene name cells should be skipped.
- `organism`: (Optional) Specify the organism for matching IDs. [human, rat, mouse, rabbit]
- `fasta`: (Optional) FASTA file to use when mode is all or fasta.
- `keep_empty`: A boolean value indicating if rows with empty reduced gene names should be retained.

#### Run Method

```{r remap_gene_names}
# Define the parameters
filtered_prot_data2 <- filtered_prot_data[filtered_prot_data$FilteredProteinIDs != "",]
mode <- "uniprot_one"  # Choose from 'all', 'fasta', 'uniprot', 'uniprot_primary', 'uniprot_one'
protein_column <- "FilteredProteinIDs"
gene_column <- "Genes"  # Optional, defaults to "Gene Names"
res_column <- "RemappedGeneNames"  # Set to NULL to overwrite 'gene_column'
skip_filled <- TRUE
organism <- "rat"  
fasta <- NULL  # Specify FASTA file path, if required
keep_empty <- TRUE

remapped_prot_results <- proharmed::remap_genenames(data =  filtered_prot_data2, mode = mode, 
                                                    protein_column = protein_column, 
                                                    gene_column = gene_column, 
                                                    res_column = res_column, 
                                                    skip_filled = skip_filled, 
                                                    organism = organism, 
                                                    fasta = fasta, 
                                                    keep_empty = keep_empty)

remapped_prot_data <- remapped_prot_results$Modified_Data
```

#### Save Dataframes

```{r remap_gene_names_save, include=FALSE}
# Create a subdirectory for remap_gene_names
if(!is.null(final_out_dir)){
  subdir <- file.path(final_out_dir, "remap_gene_names", "")
  dir.create(subdir, recursive = TRUE, showWarnings = FALSE)
  # Save the data.frames in the list to the specified subdirectory
  save_data_frames(remapped_prot_results, subdir)
}
```


#### Visualizing results

##### Overview Plot

`create_overview_plot` generates an overview plot of the logging data. This plot provides a high-level summary of the data processing steps.

Parameters:

- `logging`: A DataFrame containing the overview logging data returned by one of the proharmed methods.
- `out_dir`: (Optional) The directory where the plot will be saved.
- `file_type`: (Optional) The file format for the saved plot (e.g., png, pdf, jpg).

Returns: A list of two `ggplot` objects.

```{r remap_gene_names_plots, message=FALSE, warning=FALSE}
if(save_plots && !is.null(final_out_dir))
{ out_dir=subdir } else { out_dir=NULL }

file_type="png"

# Generate and visualize the overview plot
overview_plots <- proharmed::create_overview_plot(logging = remapped_prot_results$Overview_Log, 
                                                  out_dir = out_dir, file_type = file_type)
print(overview_plots[[1]]) # Distribution of number of entries per line
print(overview_plots[[2]]) # Number of entries added and removed

```


### Map Orthologs

The `map_orthologs` function from the `proharmed` package is designed to map gene names from one organism to their orthologs in another organism. This function is particularly useful for cross-species analyses and comparative genomics studies.

#### Parameters

- `data`: A DataFrame containing a column with gene names.
- `gene_column`: The name of the column in data that contains the gene names.
- `organism`: The organism corresponding to the current gene names in the data.
- `tar_organism`: The target organism to which you want to map the orthologs.
- `res_column`: The name of the column where the orthologs will be stored. If NULL, the gene_column will be overwritten.
- `keep_empty`: A boolean value indicating whether rows with empty cells (where orthologs could not be found) should be kept or deleted.

#### Run Method

```{r map_orthologs}
# Define the parameters
gene_column <- "ReducedGeneNames"
organism <- "rat" 
tar_organism <- "human"  #
res_column <- "Orthologs"  # Column name for the results
keep_empty <- TRUE  # Whether to keep rows with no mapped orthologs

# Mapping orthologs
orthologs_prot_results <- proharmed::map_orthologs(data = remapped_prot_results, gene_column = gene_column, 
                                                   organism = organism, tar_organism = tar_organism, 
                                                   res_column = res_column, keep_empty = keep_empty)

orthologs_prot_data <- orthologs_prot_results$Modified_Data
```

#### Save Dataframes

```{r map_orthologs_save, include=FALSE}
# Create a subdirectory for map_orthologs
if(!is.null(final_out_dir)){
  subdir <- file.path(final_out_dir, "map_orthologs", "")
  dir.create(subdir, recursive = TRUE, showWarnings = FALSE)
  # Save the data.frames in the list to the specified subdirectory
  save_data_frames(orthologs_prot_results, subdir)
}
```

#### Visualizing results

##### Overview Plot

`create_overview_plot` generates an overview plot of the logging data. This plot provides a high-level summary of the data processing steps.

Parameters:

- `logging`: A DataFrame containing the overview logging data returned by one of the proharmed methods.
- `out_dir`: (Optional) The directory where the plot will be saved.
- `file_type`: (Optional) The file format for the saved plot (e.g., png, pdf, jpg).

Returns: A list of two `ggplot` objects.

##### Detailed Plot for Reduces Gene Names

`create_ortholog_detailed_plot` is a tailored function for creating a detailed visualization of the logging data generated by the `map_orthologs` method in the proharmed package. This function produces a comprehensive plot that illustrates the intricate process of mapping orthologs from one organism to another, showcasing the critical steps and results achieved during this procedure.

Parameters

- `logging`: A DataFrame containing detailed logging data returned by the `map_orthologs` method of proharmed.
- `organism`: The source organism used in the `map_orthologs` method.
- `out_dir`: (Optional) The directory where the plot will be saved. If not specified, the plot is not saved but returned for display.
- `file_type`: (Optional) The file format for the saved plot (options include 'png', 'pdf', 'jpg', etc.).

Returns: A `ggplot` object.

```{r map_orthologs_plots, message=FALSE, warning=FALSE}
if(save_plots && !is.null(final_out_dir))
{ out_dir=subdir } else { out_dir=NULL }

file_type="png"

# Generate and visualize the overview plot
overview_plots <- proharmed::create_overview_plot(logging = orthologs_prot_results$Overview_Log, 
                                                  out_dir = out_dir, file_type = file_type)
print(overview_plots[[1]]) # Distribution of number of entries per line
print(overview_plots[[2]]) # Number of entries added and removed

# Generate and visualize the detailed reduced gene names plot [needs fix]
detailed_plot <- proharmed::create_ortholog_detailed_plot(logging = orthologs_prot_results$Detailed_Log, 
                                                          organism = organism, 
                                                          out_dir = out_dir, file_type = file_type)
print(detailed_plot) # Number of entries removed by cause
```

### Additional Overview

```{r}
# Function to categorize impact based on missing IDs
categorize_impact <- function(ids, filtered_ids) {
  ids_split <- strsplit(as.character(ids), ";")[[1]]
  filtered_ids_split <- strsplit(as.character(filtered_ids), ";")[[1]]

  missing_ids <- setdiff(ids_split, filtered_ids_split)
  if (length(missing_ids) > 0) {
    if (length(filtered_ids_split) == 0) {
      return(data.frame(id = missing_ids, impact = paste0("removed Group by ", length(missing_ids))))
    } else {
      return(data.frame(id = missing_ids, impact = paste0("reduce Group by ", length(missing_ids))))
    }
  }
}
# Apply the function to each row and combine results
impact_df <- do.call(rbind, apply(filtered_prot_data[, c("Protein.Ids", "FilteredProteinIDs")], 1, function(x) categorize_impact(x[1], x[2])))

# Count occurrences in impact_df
impact_counts <- impact_df %>%
  count(impact)

# Create the plot
ggplot(impact_counts, aes(x = impact, y = n, fill = impact)) +
  geom_bar(stat = "identity") +
  geom_text(aes(label = n), vjust = -0.5, position = position_dodge(0.9)) +
  labs(title = "Impact of Protein ID Filtering", y = "Count", x = "Impact Type") +
  theme_minimal()

```

```{r}
# Count removed groups
removed_groups <- nrow(filtered_prot_data) - nrow(remapped_prot_data)

# Count missing gene names in RM
missing_gene_names <- sum(remapped_prot_data$RemappedGeneNames == "")
missing_genes <- sum(remapped_prot_data$Genes == "")

# Prepare data for plot
plot_data <- data.frame(
  Category = c("Missing Groups\nafter Filtering", "Missing Genes", "Missing Genes Without\nRemapping"),
  Count = c(removed_groups, missing_gene_names, missing_genes)
)

# Plot for removed groups and missing gene names
ggplot(plot_data, aes(x = Category, y = Count, fill = Category)) +
  geom_bar(stat = "identity") +
  geom_text(aes(label = Count), vjust = -0.5, position = position_dodge(width = 0.9), size = 4) +
  labs(title = "Analysis of Protein Groups", y = "Count", x = "Category") +
  theme_minimal()
```


